proxy_top_restructure

NathanSegerlind opened the pull request:
refactored/restructured ProxySuspiciousConnects analysis to use a train/score flow at the top level

NathanSegerlind committed:
proxy_top_restructure

refactored/restructured ProxySuspiciousConnects analysis to use a train/score flow at the top level

NathanSegerlind commented:
not tested yet...

NathanSegerlind commented:
one thing that this does NOT try to do is separate word calculation from scoring... while it makes a lot of sense from a performance POV in the current use case ("train" the model to assign scores on the same batch of data) ,  it's extremely weird in the more normal "train" on one batch of data, and later score with different batches of data (this is what would happen in the much discussed near-real time use case). Separating word construction from scoring in a rash way could paint us in a bad spot for near real time.

rabarona commented on src/main/scala/org/opennetworkinsight/proxy/ProxySuspiciousConnectsModel.scala:
```diff
@@ -0,0 +1,192 @@
+    val agentToCountBC = sc.broadcast(agentToCount)
+
+    val udfWordCreation =
+      ProxyWordCreation.udfWordCreation(topDomains, agentToCountBC, timeCuts, entropyCuts, agentCuts)
```
It's great we are not re-calculating timeCuts, entropyCuts and agentCuts, those are some of the most expensive actions in our process. With them happening only once we are going to reduce the time significantly but I'm thinking and asking myself - since the other day- if we really need to re-calculate or re-create data with words. Essentially it should be the same data set used during the training and the scoring that's why we are carrying the rawData data frame. What if we cache the entire data frame with words instead of just returning the cuts? And that way we skip the word creation.

I know this could not make sense in case we want to break the logic - training and scoring - in the future, but for now it looks to me they are interdependent so we could take advantage of that.

There are two possible ways: 
1. registerTempTable
2. persist

rabarona commented:
Overall looks good, just want to understand better the need/use of object and class ProxySuspiciousConnectsModel.

NathanSegerlind commented:
I here you about not recalculating the words..... building the model and scoring on one batch of data is probably going to remain a very common use case

but... I recognized the complication with splitting training from scoring, and decided that this was a point to say "stop" for this check-in so we could think more about have both the train/score on one batch and train on one batch and score on another both work cleanly

NathanSegerlind committed:
proxy_top_restructure

corrected schema mismatch and incorrect filter logic (< becomes <= when applying threshold)

NathanSegerlind committed:
proxy_top_restructure

Corrected a serialization error by moving the  proxy scoring function to a serializable class.
